### PR TITLE
COMMERCE-10498 Discontinued skus are not replaced using Quick Add to Cart

### PR DESCRIPTION
**Pull request by rosdb**
Supersedes [#3182](https://github.com/liferay-frontend/liferay-portal/pull/3182) and [#3850](https://github.com/liferay-commerce/liferay-portal/pull/3850)

Task: [COMMERCE-10498](https://issues.liferay.com/browse/COMMERCE-10498) and [COMMERCE-11205](https://liferay.atlassian.net/browse/COMMERCE-11205)
Sub-Task: [COMMERCE-11546](https://issues.liferay.com/browse/COMMERCE-11546) and [COMMERCE-11375](https://liferay.atlassian.net/browse/COMMERCE-11375)

### CODE REVIEW

**Comment by rosdb** [comment]
ci:test:sf

**Comment by rosdb** [comment]
ci:test:relevant

**Comment by rosdb** [comment]
ci:test:relevant

**Comment by rosdb** [comment]
ci:test:relevant

**Comment by rosdb** [comment]
ci:test:relevant

**Comment by rosdb** [comment]
ci:test:relevant

**Comment by rosdb** [comment]
ci:test:relevant

**Comment by rosdb** [comment]
ci:stop

**Comment by rosdb** [comment]
ci:test:relevant

**Comment by rosdb** [comment]
ci:stop

**Comment by rosdb** [comment]
ci:test:relevant

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/CartQuickAdd.js [hunk 1]
```diff
@@ -158,37 +128,46 @@ export default function CartQuickAdd() {
 	const handleAddToCartClick = () => {
 		const readySKUs = selectedSKUs.map((selectedSKU) => {
 			const {
+				discontinued,
 				id,
 				productConfiguration,
-				replacementSkuId,
+				replacementSku,
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/CartQuickAdd.js:134]: replacementSKU

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/CartQuickAdd.js [hunk 2]
```diff
@@ -158,37 +128,46 @@ export default function CartQuickAdd() {
...
 				sku,
 				urls,
 			} = selectedSKU;
 
-			if (replacementSkuId) {
-				const replacementSKU = availableSKUs.find(
-					(availableSKU) => availableSKU.id === replacementSkuId
-				);
+			if (discontinued) {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/CartQuickAdd.js:139]: Check if selectedSKU is unavailable and if replacementSKU exists

Comment (alessiorendina) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/CartQuickAdd.js:139]: `if ((selectedSKU.availability?.label !== 'available') && replacementSKU)`

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/CartQuickAdd.js [hunk 3]
```diff
@@ -158,37 +128,46 @@ export default function CartQuickAdd() {
...
 					),
-					replacementSkuId: id,
-					settings: productConfiguration,
-					skuId: replacementSKU.id,
-					urls: replacementSKU.urls,
+					replacedSkuId: id,
+					settings: replacementConfiguration,
+					skuId,
+					skuOptions,
+					urls,
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/CartQuickAdd.js]:
```
return {
					...replacementSKU,
					quantity: getCorrectedQuantity(
						replacementConfiguration,
						sku,
						cartItems
					),
					replacedSkuId: id,
					settings: replacementConfiguration,
```

**Review by alessiorendina — CHANGES_REQUESTED** [review]
Hey @rosdb please check my comments, thank you!

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/CartQuickAdd.js (hunk 1 above)
Comment (rosdb) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/CartQuickAdd.js]: ✅

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/CartQuickAdd.js (hunk 2 above)
Comment (rosdb) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/CartQuickAdd.js]: ✅

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/CartQuickAdd.js (hunk 3 above)
Comment (rosdb) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/CartQuickAdd.js]: @alessiorendina, there are side effects to calling `addToCart` method, so we need, from `replacementSKU` object, also the `price` and the `urls` properties. In the last commit, I hardcoded it to don't push breakable code.

File: modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties [hunk 4]
```diff
@@ -14053,6 +14053,7 @@ replace-files=Replace Files
 replace-with=Replace with:
 replacement=Replacement
 replacement-product=Replacement Product
+replacement-product-for-x=Replacement Product for {0}.
```
Comment (alessiorendina) [inline modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties:14056]: This one should be moved in another commit

**Review by alessiorendina — CHANGES_REQUESTED** [review]
Hey @rosdb it looks good to me, just one comment. Please move the commits that contain auto generate files at the end. Like buildREST and buildLang

**Comment by alessiorendina** [comment]
ci:test:sf

**Comment by rosdb** [comment]
ci:test:sf

**Comment by rosdb** [comment]
ci:test:sf

**Comment by rosdb** [comment]
ci:test:sf

**Comment by rosdb** [comment]
ci:test:relevant

**Comment by rosdb** [comment]
ci:test:relevant

**Comment by rosdb** [comment]
ci:stop

**Comment by rosdb** [comment]
ci:test:relevant

**Comment by rosdb** [comment]
ci:test:relevant

**Review by alessiorendina — CHANGES_REQUESTED** [review]
@rosdb LGTM, just the commit order. Can you move autogenerated at the end of the history?

**Comment by rosdb** [comment]
> @rosdb LGTM, just the commit order. Can you move autogenerated at the end of the history?
> 
> buildLang and buildREST

@alessiorendina, done ✅

**Review by alessiorendina — APPROVED** [review]

**Comment by victoroscuro** [comment]
ci:test:commerce-functional

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/MiniCart.js [hunk 5]
```diff
@@ -119,14 +121,29 @@ function MiniCart({
...
 			}
 		},
 		[onAddToCart]
 	);
 
+	const updateReplacedSKUList = useCallback(
+		() =>
+			cartState.cartItems
+				? setReplacementSKUList(
+						cartState.cartItems.filter((i) =>
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/MiniCart.js:135]: i -> cartItem

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/CartItemsList.js [hunk 6]
```diff
@@ -26,20 +26,37 @@ export default function CartItemsList({showPriceOnApplicationInfo = false}) {
 		cartState,
 		isUpdating,
 		labels,
+		replacementSKUList,
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/CartItemsList.js:29]: Why capital letters for the Sku? Let's keep it consistent with other names, either change this or the other ones

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/MiniCart.js [hunk 7]
```diff
@@ -119,14 +121,29 @@ function MiniCart({
 				});
 
 				onAddToCart(latestActionURLs, latestCartState);
-			}
-			catch (error) {
+			} catch (error) {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/MiniCart.js:124]: Did you run the source formatter in this module? I think SF won't like it

**Review by alessiorendina — CHANGES_REQUESTED** [review]
@rosdb LGTM overall, just SF stuff. Thank you!

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/MiniCart.js (hunk 5 above)
Comment (rosdb) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/MiniCart.js]: ✅

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/CartItemsList.js (hunk 6 above)
Comment (rosdb) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/CartItemsList.js:29]: > https://github.com/liferay-commerce/liferay-portal/pull/3859#discussion_r1239795454 😄 
done, I kept the name used in the client

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/MiniCart.js (hunk 7 above)
Comment (rosdb) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/MiniCart.js]: ✅

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/CartItemsList.js (hunk 6 above)
Comment (rosdb) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/CartItemsList.js:29]: Those are what the server expects, and I can't modify them. @alessiorendina

**Comment by 22LucaGiannotta22** [comment]
ci:test:relevant

**Comment by 22LucaGiannotta22** [comment]
ci:test:commerce-functional
